server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java

```diff
@@ -19,39 +19,178 @@
 
 package org.apache.james.webadmin.dto;
 
+import java.io.IOException;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.Iterator;
 import java.util.List;
+import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
 
+import javax.mail.Header;
 import javax.mail.MessagingException;
 
 import org.apache.james.core.MailAddress;
 import org.apache.mailet.Mail;
+import org.apache.mailet.PerRecipientHeaders;
 
 import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.Multimap;
 
 public class MailDto {
-    public static MailDto fromMail(Mail mail) throws MessagingException {
+    public static MailDto fromMail(Mail mail, List<AdditionalFields> additionalFields) throws MessagingException, MissingRequestedField {
         return new MailDto(mail.getName(),
             Optional.ofNullable(mail.getSender()).map(MailAddress::asString),
             mail.getRecipients().stream().map(MailAddress::asString).collect(Guavate.toImmutableList()),
             Optional.ofNullable(mail.getErrorMessage()),
-            Optional.ofNullable(mail.getState()));
+            Optional.ofNullable(mail.getState()),
+            Optional.ofNullable(mail.getRemoteHost()),
+            Optional.ofNullable(mail.getRemoteAddr()),
+            Optional.ofNullable(mail.getLastUpdated()),
+            fetchAttributes(additionalFields, mail),
+            fetchPerRecipientsHeaders(additionalFields, mail),
+            fetchHeaders(additionalFields, mail),
+            fetchBody(additionalFields, mail),
+            fetchMessageSize(additionalFields, mail));
+    }
+
+    private static Optional<Long> fetchMessageSize(List<AdditionalFields> additionalFields, Mail mail) throws MissingRequestedField {
+        if (!additionalFields.contains(AdditionalFields.MESSAGE_SIZE)) {
+            return Optional.empty();
+        }
+        try {
+            return Optional.ofNullable(new Long(mail.getMessageSize()));
+        } catch (MessagingException e) {
+            throw new MissingRequestedField("messageSize");
+        }
+    }
+
+    private static Optional<String> fetchBody(List<AdditionalFields> additionalFields, Mail mail) throws MissingRequestedField {
+        if (!additionalFields.contains(AdditionalFields.BODY)) {
+            return Optional.empty();
+        }
+        try {
+            return Optional.ofNullable(mail.getMessage().getContent().toString());
```

Comment:
we try as much as possible to not rely on toString behavior.
In Java, toString is meant for displaying debug information, not to have production code rely on it.

Reply:
It has an issue, it dumps everything about the mail:
`"body": "Date: Wed, 6 Jun 2018 15:54:10 +0200 (CEST)\r\nMessage-ID: <1866768727.0.1528293250828@gautier-HP-ProBook-430-G4>\r\nMIME-Version: 1.0\r\nContent-Type: text/plain; charset=UTF-8\r\nContent-Transfer-Encoding: 7bit\r\nheaderName4: value6\r\nheaderName4: value7\r\nheaderName3: value5\r\nheaderName3: value8\r\n\r\nMy awesome body!!",`

Reply:
Hmm maybe in the end we don't really care about the `body` which is not readabe anyway (encoding, mime parts etc...). And if needed the eml can be downloaded (cf antoine's task).

Maybe we should more aim for a `textBody` & `htmlBody` here...

You can use `MessageContentExtractor` for getting these information.

(You'll need to MIME4J parse the message before. Please read MessageFactory::parse for this)

What do you think about this other team members?